compiler.rst

```diff
@@ -159,49 +159,88 @@ are very rare.  However, if you've allocated a PyObject, you must tell
-supporting functions, and connecting them as needed.
-
-Do realize that there is no automated nor symbolic connection between
-the grammar specification and the nodes in the parse tree.  No help is
-directly provided by the parse tree as in yacc.
-
-For instance, one must keep track of which node in the parse tree
-one is working with (e.g., if you are working with an 'if' statement
-you need to watch out for the ':' token to find the end of the conditional).
-
-The functions called to generate AST nodes from the parse tree all have
-the name :samp:`ast_for_{xx}` where *xx* is the grammar rule that the function
-handles (``alias_for_import_name`` is the exception to this).  These in turn
-call the constructor functions as defined by the ASDL grammar and
-contained in :file:`Python/Python-ast.c` (which was generated by
-:file:`Parser/asdl_c.py`) to create the nodes of the AST.  This all leads to a
+Source Code to AST
+------------------
+
+The AST is generated from source code using the function
+``_PyParser_ASTFromString()`` and ``_PyParser_ASTFromFile()``
+(from :file:`Parser/peg_api.c`) depending on the input type.
+
+After some checks, a helper function in :file:`Parser/parser.c` begins applying
+production rules on the source code it receives; converting source code to
+tokens and matching these tokens recursively to their corresponding rule.  The
+rule's corresponding rule function is called on every match.  These rule
+functions follow the format :samp:`xx_rule`.  Where *xx* is the grammar rule
+that the function handles and is automatically derived from
+:file:`Grammar/python.gram` by :file:`Tools/peg_generator/pegen/c_generator.py`.
+
+Each rule function in turn creates an AST node as it goes along.  It does this
+by allocating all the new nodes it needs, calling the proper AST node creation
+functions for any required supporting functions and connecting them as needed.
+This continues until all nonterminal symbols are replaced with terminals.  If an
+error occurs, the rule functions backtrack and try another rule function.  If
+there are no more rules, an error is set and the parsing ends.
+
+The AST node creation functions have the name :samp:`_PyAST_{xx}` where *xx* is
+the AST node that the function creates. These are defined by the ASDL grammar
+and contained in :file:`Python/Python-ast.c` (which is generated by
+:file:`Parser/asdl_c.py` from :file:`Parser/Python.asdl`).  This all leads to a
 sequence of AST nodes stored in ``asdl_seq`` structs.
 
-Function and macros for creating and using ``asdl_seq *`` types as found
-in :file:`Python/asdl.c` and :file:`Include/asdl.h` are as follows:
-
-``_Py_asdl_seq_new(Py_ssize_t, PyArena *)``
-        Allocate memory for an ``asdl_seq`` for the specified length
-``asdl_seq_GET(asdl_seq *, int)``
+To avoid bad performance during backtracking, rules (other than the top level
+ones) are memoized.  Each rule function checks if a memoized version exists and
```

Comment:
```suggestion
To avoid bad performance during backtracking, some rules, chosen by applying a
flag in the grammar file, are memoized.  Each rule function checks if a memoized
version exists and
```

Reply:
Wow thanks! I didn't know that a flag existed. Cool, I just noticed the ``(memo)`` flag in the grammar file.